databooks/affirm.py

```diff
@@ -0,0 +1,244 @@
+"""Functions to safely evaluate strings and inspect notebook."""
+import ast
+from copy import deepcopy
+from functools import reduce
+from itertools import compress
+from pathlib import Path
+from typing import Any, Callable, Dict, Iterable, List, Tuple
+
+from databooks import JupyterNotebook
+from databooks.data_models.base import DatabooksBase
+from databooks.logging import get_logger, set_verbose
+
+logger = get_logger(__file__)
+
+_ALLOWED_BUILTINS = (
+    all,
+    any,
+    enumerate,
+    filter,
+    getattr,
+    hasattr,
+    len,
+    list,
+    range,
+    sorted,
+)
+_ALLOWED_NODES = (
+    ast.Add,
+    ast.And,
+    ast.BinOp,
+    ast.BitAnd,
+    ast.BitOr,
+    ast.BitXor,
+    ast.BoolOp,
+    ast.boolop,
+    ast.cmpop,
+    ast.Compare,
+    ast.comprehension,
+    ast.Constant,
+    ast.Dict,
+    ast.Eq,
+    ast.Expr,
+    ast.expr,
+    ast.expr_context,
+    ast.Expression,
+    ast.For,
+    ast.Gt,
+    ast.GtE,
+    ast.In,
+    ast.Is,
+    ast.IsNot,
+    ast.List,
+    ast.ListComp,
+    ast.Load,
+    ast.LShift,
+    ast.Lt,
+    ast.LtE,
+    ast.Mod,
+    ast.Name,
+    ast.Not,
+    ast.NotEq,
+    ast.NotIn,
+    ast.Num,
+    ast.operator,
+    ast.Or,
+    ast.RShift,
+    ast.Set,
+    ast.Slice,
+    ast.slice,
+    ast.Str,
+    ast.Sub,
+    ast.Tuple,
+    ast.UAdd,
+    ast.UnaryOp,
+    ast.unaryop,
+    ast.USub,
+)
+
+
+class DatabooksParser(ast.NodeVisitor):
+    """AST parser that disallows unsafe nodes/values."""
+
+    def __init__(self, **variables: Any) -> None:
```

Comment:
Why do you not specify the variables?

Reply:
I wanted this to be more general. The contract is: give me variables (can be constants or pydantic models) and I'll tell you if this is ok to run